_drafts/reading-legacy-code-to-harmonize-ruby-and-rails-code-loaders.md

```diff
@@ -0,0 +1,701 @@
+
+![`auxiliary/pro/social_engineering/web_phish` `Metasploit3#run`](/images/auxiliary-pro-social-engineering-web-phish-run.png)
+
+This leaves the data flow for `nesting` being `name` → `name.presence` → `klass_name` → `klass_name.to_s` → `klass_name.to_s.scan` → `nesting`.  The original messed up `namespace`, which was `#<Module:0x0000010c916448>` must either be either the direct `Module#name` if `Module#name` contains no `::` or be part of `Module#name`.
+
+### Debugging
+
+I need to look at `name`, so I'm going to debug the call.  Normally, I'd use Rubymine's remote debugger, but at this time, Metasploit Pro had an issue where it loaded symlink paths instead of absolute paths, so ruby-debug-ide couldn't register correct breakpoints, and pry and debugger aren't available in the gemset that msfpro will use, so I'm stuck with print debugging.
+
+{% highlight sh %}
+vim ~/.rvm/gems/ruby-1.9.3-p551@pro/gems/activesupport-3.2.2/lib/active_support/dependencies.rb
+{% endhighlight %}
+
+![`ModuleConstMissing#const_missing` debugging](/images/module-const-missing-const-missing-debugging.png)
+
+Let's retry the reproduction steps:
+
+{% highlight sh %}
+cd ~/git/rapid7/pro/engine
+RAILS_ENV=development ./msfpro
+{% endhighlight %}
+
+{% highlight irb %}
+msf > irb
+>> Object.send(:remove_const, :SocialEngineering)
+>> exit
+msf > use auxiliary/pro/social_engineering/web_phish
+msf > set CAMPAIGN_ID 1
+msf > run
+{% endhighlight %}
+
+![debugging output](/images/debugging-output.png)
+
+Ok, so it's a namespace Module for `Metasploit3` and `Metasploit3` happens to be the name of the class in `auxiliary/pro/social_engineering/web_phish`.
+
+![`auxiliary/pro/social_engineering/web_phish` `Metasploit3` class](/images/auxiliary-pro-social-engineering-web-phish-class.png)
+
+The `MetasploitN` class naming convention is used to indicate the minimum major version of Metasploit Framework required to run the module, so there are 100s of `Metasploit3` and `Metasploit4` classes.
+
+### Tracking anonymous Module creation
+
+So, however `use` loads the Metasploit Module, it must introduce an anonymous Module (`Module.new`) that is used to prevent the `Metasploit3` classes from colliding with each other.  So, let's look for `Module.new` in `metasploit-framework/lib`.
+
+![`Module.new` in `metasploit-framework/lib`](/images/module-new-in-metasploit-framework-lib.png)
+
+If I was declaring an anonymous `Module` to wrap something, `wrap` would be a good name, so let's look at each of those.
```

Comment:
If I **~~was~~** **were** declaring an anonymous `Module` to wrap something, `wrap` would be a good name, so let's look at each of those.